iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,164 @@
+        {
+            return cxx::error<ServiceRegistryError>(ServiceRegistryError::SERVICE_DESCRIPTION_ALREADY_ADDED);
+        }
+    }
+
+    if (!m_serviceDescriptionVector.push_back(serviceDescription))
+    {
+        return cxx::error<ServiceRegistryError>(ServiceRegistryError::SERVICE_REGISTRY_FULL);
+    }
+    m_serviceMap.insert({serviceDescription.getServiceIDString(), m_serviceDescriptionVector.size() - 1});
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), m_serviceDescriptionVector.size() - 1});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const CaproIdString_t& service, const CaproIdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription)
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    bool removedElement{false};
+
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        if (m_serviceDescriptionVector[index] == serviceDescription)
+        {
+            uint64_t removedValue{0U};
+            bool removedEntry{false};
+            for (auto it = m_serviceMap.begin(); it != m_serviceMap.end();)
+            {
+                if (it->second == index)
+                {
+                    removedValue = it->second;
+                    it = m_serviceMap.erase(it);
+                    removedEntry = true;
+                    continue;
+                }
+                else if (removedEntry && it->second > removedValue)
+                {
+                    // update index due to removed element
+                    it->second--;
+                }
+                it++;
+            }
+
+            removedValue = 0U;
+            removedEntry = false;
+            for (auto it = m_instanceMap.begin(); it != m_instanceMap.end();)
+            {
+                if (it->second == index)
+                {
+                    removedValue = it->second;
+                    it = m_instanceMap.erase(it);
+                    removedEntry = true;
+                    continue;
+                }
+                else if (removedEntry && it->second > removedValue)
+                {
+                    // update index due to removed element
+                    it->second--;
+                }
+                it++;
+            }
+            m_serviceDescriptionVector.erase(iterator);
+            removedElement = true;
+            // There can be not more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const CaproIdString_t& service,
                            const CaproIdString_t& instance) const
 {
-    if (instance == iox::cxx::string<100>(capro::AnyInstanceString))
```

Comment:
Please refactor find as well with the suggested data structure: `std::map<std::map<CaproIdString_t, std::set<CaproIdString_t>>, CaproIdString_t> m_services`.

Then you can write the algorithm with much less code, somehow like this:
```cpp
cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> searchResult;

auto serviceStartIter = (service == capro::AnyServiceString) ? m_service.begin() : m_services.find(service);
auto serviceEndIter = (service == capro::AnyServiceString) ? m_service.end() : serviceStartIter + 1;

for(auto iter = serviceStartIter; iter != m_service.end() && iter != serviceEndIter; ++iter) {
  auto instanceStartIter = (service == capro::AnyInstanceString) ? iter.begin() : iter.find(service);
  auto instanceEndIter = (service == capro::AnyInstanceString) ? iter.end() : instanceStartIter + 1;
  for(auto iter2 = instanceStartIter; iter2 != iter.end() && iter2 != instanceEndIter; ++iter2) {
     for(auto event : iter2 )
       searchResult.emplace_back(iter.first, iter2.first, event);
  }
}
// maybe you have to rename iter to serviceIter, iter2 to instanceIter but this could would be much cleaner and faster
```

Reply:
I need to look a at the current implementation but I am sure that any solution with maps and sets is inferior to prefix tree implementations I propose (and the latter is easier to implement than a map or set).

The reason is simple: is completing most operations in O(#length of string to insert), as opposed to anything with maps which at least need O(log(#elements in container)*#length of keys to compare). These structures are much harder to implement with the logarithmic guarantees as well, compared to linear in the key length. This is especially true for the multimaps needed here. 

I have not checked the correctness of the above code, but if it is correct the estimate still holds true and gets worse with any iteration done above.

Reply:
@elfenpiff That's a neat and small implementation of `find()`! However, as stated in the other threads I'm not convinced that we will be able to build a fixed size `cxx::map`, that can be place in shared memory. What is your take on that?

Reply:
This should be feasable. I see a map implementation as a challenge but not the static memory part and in some previous company I already have seen such a map. The tree can be implemented in the same fashion like the linked list which we already have (from the memory point of view).

Reply:
This is true, it is feasible with the same approach as the list, but it would still be the wrong structure for the job due to this nesting of maps and sets. 

And it is a challenge I intend to tackle in the near future when we have a working approach. We would even be able to compare efficiency then, and my bet is against any map based approach (we can try with `std::map`, which is likely better optimized then what we will be able to come up without serious time investment).